### ISSUE TITLE
getState: Support csv and txt formats for `chifra state --call <address>!<function> <blocks>`

### ISSUE THREAD

**Issue by mds1** [TrueBlocks/trueblocks-core#1690]
Currently CSV and TXT format options are ignored when used with this command, but it would be great to support these. I think there are a few ways to handle the output depending on the scenario:

1. One output, with ABI: fairly straightforward, just have an "output" column
2. Multiple outputs, with ABI: one column per output, and default names of "output1", "output2", etc. is ideal. If it's easier, even a single output with results concatenated into a string works since it isn't much work to post-process
3. One output, no ABI: a single "output" column with the response, of course not decoded in this case
4. Multiple outputs, no ABI: the same as 3, and the user will be responsible for decoding

**Comment by tjayrush** [comment]
Do you know of any contracts with functions that have multiple return values? Can you point me to an example?

I think we can do the same thing we do with `compressedTx` output from `chifra export` --> basically, we form a C++ like function call with the values in a single column so it works for both one return and many. It's easy to parse.

**Comment by mds1** [comment]
Compound's Comptroller (0x3d9819210A31b4961b30EF54bE2aeD79B9c9Cd3B) has a method with the function signature `function getAccountLiquidity(address account) public view returns (uint, uint, uint)` which seems like a good test. You can call it with an input account of 0x57F0a2944212e193e2E1fAFA625E31D24FBe2209, which is a random address I found that should return something like (0, 12960375376800454452137, 0). The first value is zero on success, and only one of the second and third values should be non-zero

(In case it matters, that Comptroller is a proxy contract, and the implementation it delegates to is 0xBafE01ff935C7305907c33BF824352eE5979B526)

**Comment by tjayrush** [comment]
I think I understand this problem.

This is a proxy contract, but it doesn't implement what I've been calling the 'semi-standard' proxy routine called `implementation()` which has four-byte signature of `0x5c60da1b`.

This contract has `comptrollerImplementation` with a four-byte of `0xbb82aa5e`.

If I call that function `0xbb82aa5e` against the original address `0x3d9819210A31b4961b30EF54bE2aeD79B9c9Cd3B` I get this `0xbafe01ff935c7305907c33bf824352ee5979b526`, which is the proxy you mention above.

It's probably not a good idea to add that as a special case because where does that stop? Each new version of the "implementation" for each new proxy would make the code un-maintainable.

I think instead what I should do is enable a command line option for `chifra state --call` called `--proxy`. If that value is present, then we would use it as the source of both the ABI and the result you're looking for.

But I'm not sure that's the best solution either.

May I ask you how `seth` handles this case? Does it find the `proxy` automatically or do you have to know that already?

**Comment by tjayrush** [comment]
We still get the wrong answer, though, even when I use the proxy address directly which may be a different problem.

**Comment by mds1** [comment]
> I think instead what I should do is enable a command line option for `chifra state --call` called `--proxy`. If that value is present, then we would use it as the source of both the ABI and the result you're looking for.

Hm, not sure I follow exactly how this proxy flag would work?

------------------------

Can you help me understand why it matters whether or not we call a proxy? If we provide the function signature / calldata (which we do with `chifra state --call`), then it shouldn't matter whether the target address is a proxy or not, because the contract handles the delegation and provides the return data. 

That's how it works with seth—it knows nothing about whether or not the target is a proxy and just executes an `eth_call` based on the provided data:

```sh
seth call 0x3d9819210A31b4961b30EF54bE2aeD79B9c9Cd3B "getAccountLiquidity(address)(uint,uint,uint)" 0x57F0a2944212e193e2E1fAFA625E31D24FBe2209
# prints:
#   0
#   12948041749860972653620
#   0

**Comment by tjayrush** [comment]
I'm not sure how Seth works but for us the ABI of the proxy does not include the function being called into the underlying contract. In this case, this is the ABI for the `0x3d...` contract: https://etherscan.io/address/0x3d9819210A31b4961b30EF54bE2aeD79B9c9Cd3B#readContract.

The only way we can get to the underlying contract is to follow the `comptrollerImplementation` end point, but that's not a standard way to describe the `implementation`. Most other `proxies` have an `implementation` end point.

If we query against the result of querying against `comprtollerImplementation` then we get the ABI we need.

Without the right ABI, we can't decode the output.

My guess (and again, I don't know how seth does it) is that seth is doing something similar.

In any case, see my next comment.

**Comment by tjayrush** [comment]
We extended the `chifra state --call` output to

1) display multiple outputs if they are present
2) allow for `--fmt txt` and `--fmt csv` output

Closing this but opening a new issue to add an option for `--implementation <address>` which, if present will call this contract as well as the topmost one.

https://github.com/TrueBlocks/trueblocks-core/issues/1845

**Comment by tjayrush** [comment]
Here's an interesting fact. I can see that the implementation has changed on this contract:

```
chifra state  --call 0x3d9819210A31b4961b30EF54bE2aeD79B9c9Cd3B!0xbb82aa5e 10092000 --fmt csv
```

(a query at an old block) returns

```
"blocknumber","address","signature","encoding","bytes","compressedresult"
"10092000","0x3d9819210A31b4961b30EF54bE2aeD79B9c9Cd3B","comptrollerImplementation()","0xbb82aa5e","","0x97bd4cc841fc999194174cd1803c543247a014fe"
```

But

```
chifra state  --call 0x3d9819210A31b4961b30EF54bE2aeD79B9c9Cd3B!0xbb82aa5e --fmt csv
```

(a query now), returns:

```
"blocknumber","address","signature","encoding","bytes","compressedresult"
"13492453","0x3d9819210A31b4961b30EF54bE2aeD79B9c9Cd3B","comptrollerImplementation()","0xbb82aa5e","","0xbafe01ff935c7305907c33bf824352ee5979b526"
```

A different value for the implementation.

**Comment by mds1** [comment]
I know you opened a new issue, but for the sake of readability I'm just posting my reply here.

> Here's an interesting fact. I can see that the implementation has changed on this contract:

Correct, this implementation is controller by Compound governance and can be changed by governance proposals. You can see the most recent proposal that changed the implementation here: https://compound.finance/governance/proposals/65

------------------------

Regarding how seth works with proxies, seth doesn't know _anything_ about the contract when you do a `seth call`. It just trusts that you do, and trusts that you gave it a valid ABI. This means it also knows nothing about the xistence of proxies.

We only need to know details about the proxy pattern used if we are trying to fetch the ABI dynamically at runtime via Etherscan (which I know TrueBlocks does in some place), but with `chifra state --call`, we are provided the ABI, ~~so I'm still not quite sure where the proxy pattern becomes relevant~~ so the proxy pattern is only relevant for decoding the output (see next comment with tentative idea—sorry, it's early and I misread your comment at first 😛)

When you execute a `seth call` to a contract with a proxy, the flow looks like this. We'll use this as our example:
```sh
comptrollerProxy=0x3d9819210A31b4961b30EF54bE2aeD79B9c9Cd3B
comptrollerImplementation= 0xBafE01ff935C7305907c33BF824352eE5979B526
user=0x57F0a2944212e193e2E1fAFA625E31D24FBe2209
seth call $comptrollerProxy "getAccountLiquidity(address)(uint,uint,uint)" $user
```

What happens is:
1. `seth` converts that call into an `eth_call` on the node with `to=$comptrollerProxy` and `data=$(seth calldata "getAccountLiquidity(address)(uint,uint,uint)" $user)`, which gives `data=0x5ec88c7900000000000000000000000057f0a2944212e193e2e1fafa625e31d24fbe2209`
2. The node calls that data on the comptrollerProxy, which iterates through it's function selectors. It compares the function selector (`0x5ec88c79`) against each function selector from it's own ABI. Because it's a proxy, the ABI has has 4 proxy-related methods, none of which are called `getAccountLiquidity`. 
3. Therefore we reach the fallback method, which says "if none of our function selectors match our ABI, try making the same call against the comptrollerImplementation instead"
4. The proxy delegatecalls to its implementation using the same `data`, and this time we do find a matching ABI and get the return data

If we had tried using a function signature that doesn't exist, the call would error:

```sh
seth call $comptrollerProxy "getAccountLiquidityyyyy(address)(uint,uint,uint)" $user
# prints:
#   seth-rpc: {"id":1,"jsonrpc":"2.0","method":"eth_call","params":[{"data":"0xbc809fab00000000000000000000000057f0a2944212e193e2e1fafa625e31d24fbe2209","to":"0x3d9819210A31b4961b30EF54bE2aeD79B9c9Cd3B"},"latest"]}
#   seth-rpc: error:   code       -32000
#   seth-rpc: error:   message    execution reverted
```

**Comment by mds1** [comment]
I see what you mean about needing the ABI to decode the output though, so one alternative to a `--proxy` flag would be to add a `--output-sig` flag that tells you how to decode it.

```
chifra state --call <address>!<function> <blocks> --output-sig "(uint,uint,uint)"
```

This is nice because (1) it's faster since you don't need a call to Etherscan, (2) it's nice not relying on Etherscan, and (3) trueblocks doesn't need to know anything about the ABI and the user can provide all required data

**Comment by tjayrush** [comment]
Rephrasing what you said, just so I can see if I understand. Because you're providing the full function signature to `seth`, it has all the information it needs to build the call data and then also decode the returned output. That makes sense to me. I can see that if `chifra` had the full signature it could do that.

Then the only other issue is the 'falling down to the fallback method` (which, if I understand you mean to say the literally fallback of the Solidity code) and the $comptrollerProxy smart contract itself handles calling into the underlying Implementation.

That should just plain work with `chifra` because I'm sure we don't do anything different than `seth` in that sense. We just call into the RPC with the `calldata`.

So, maybe my problem is that because I don't have the ABI and because I'm looking for the ABI from EtherScan (which doesn't have it for the contract I'm calling - rightly so), it fails.

That all makes perfect sense. And I like the way it works because it works for any `proxy` at all no matter how they implement themselves.

Let me see if I can't figure it out. Your `output-sig` could just be the full signature like `seth` I think as well.

Cheers. Thanks for the info.

**Comment by mds1** [comment]
> Rephrasing what you said, just so I can see if I understand. Because you're providing the full function signature to seth, it has all the information it needs to build the call data and then also decode the returned output. That makes sense to me. I can see that if chifra had the full signature it could do that.

Yep that is correct!

> Then the only other issue is the 'falling down to the fallback method` (which, if I understand you mean to say the literally fallback of the Solidity code) and the $comptrollerProxy smart contract itself handles calling into the underlying Implementation.
> 
> That should just plain work with chifra because I'm sure we don't do anything different than seth in that sense. We just call into the RPC with the calldata.

Also correct—chifra shouldn't need to worry about the fact that the proxy has a fallback method which delegates to the implementation. Those are arguably just implementation details at the contract level which, when providing the full function signature/return data format, get abstracted away by the `eth_call` RPC execution

Just for fun, I'll also mention there's a "[Beacon](https://docs.openzeppelin.com/contracts/3.x/api/proxy)" contract pattern that was developed by Dharma where you have a proxy which points to a beacon contract, which points to the implementation. This lets you deploy contract wallets for all your users, and upgrade all their wallets at once. But good luck trying to generalize some pattern like that into chifra! That's why I think (we agree) that allowing the user to specify the full function signature with return data would be a nice improvement

> Let me see if I can't figure it out. Your output-sig could just be the full signature like seth I think as well.

That's true, that would work also!

**Comment by tjayrush** [comment]
So I went with supporting the full signature. It probably isn't perfect (won't parse everything), but it works for simple signatures. I wrote about it here: https://tjayrush.medium.com/calling-smart-contracts-with-chifra-state-call-ea03b8d35ea7 trying to explain it.

Once we move to the Go version of the code we can use the `go-ethereum` package which will give us perfect parsing for both encoding and decoding.
